File: Typecast/TCTTCHeader.m
//
//  TCTTCHeader.m
//  Type Designer
//
//  Created by David Schweinsberg on 29/12/12.
//  Copyright (c) 2012 David Schweinsberg. All rights reserved.
//

#import "TCTTCHeader.h"
#import "TCDataInput.h"

@implementation TCTTCHeader

+ (BOOL)isTTCDataInput:(TCDataInput *)dataInput
{
    const int ttcf = 0x74746366;
    int ttcTag = [dataInput readInt];
    return ttcTag == ttcf;
}

- (id)initWithDataInput:(TCDataInput *)dataInput
{
    self = [super init];
    if (self)
    {
        _ttcTag = [dataInput readInt];
        _version = [dataInput readInt];
        _directoryCount = [dataInput readInt];
        NSMutableArray *tableDirectory = [[NSMutableArray alloc] initWithCapacity:_directoryCount];
        for (int i = 0; i < _directoryCount; ++i)
            [tableDirectory addObject:[NSNumber numberWithInt:[dataInput readInt]]];
        _tableDirectory = tableDirectory;
        if (_version == 0x00010000)
            _dsigTag = [dataInput readInt];
        _dsigLength = [dataInput readInt];
        _dsigOffset = [dataInput readInt];
    }
    return self;
}

@end

File: Typecast/TCTTCHeader.swift
//
//  TCTTCHeader.swift
//  Type Designer
//
//  Created by David Schweinsberg on 8/2/17.
//  Copyright © 2017 David Schweinsberg. All rights reserved.
//

import Foundation
import IOUtils

class TCTTCHeader {
  let ttcTag: Int32
  let version: Int32
  let directoryCount: Int32
  var tableDirectory: [Int32]
  let dsigTag: Int32
  let dsigLength: Int32
  let dsigOffset: Int32

  class func isTTC(data: Data) -> Bool {
    let dataInput = TCDataInput(data: data)
    let ttcf: Int32 = 0x74746366
    let ttcTag = dataInput.readInt32()
    return ttcTag == ttcf
  }

  init(data: Data) {
    let dataInput = TCDataInput(data: data)
    ttcTag = dataInput.readInt32()
    version = dataInput.readInt32()
    directoryCount = dataInput.readInt32()
    tableDirectory = []
    for _ in 0..<directoryCount {
      tableDirectory.append(dataInput.readInt32())
    }
    if version == 0x00020000 {
      dsigTag = dataInput.readInt32()
      dsigLength = dataInput.readInt32()
      dsigOffset = dataInput.readInt32()
    } else {
      dsigTag = 0
      dsigLength = 0
      dsigOffset = 0
    }
  }
}
